Ref #7860. Olimex + Freertos tutorial (#114)
micro-ROS/micro-ROS.github.io: _docs/tutorials/basic/getting_started_olimex_freertos/index.md

mkdir my_brand_new_app
cd my_brand_new_app
touch app.c app-colcon.meta
```

For this example we are going to create a ping pong app where a node sends a ping package with a unique identifier using a publisher and receives the same package a pong subscriber. The node will also answer to 

![pingpong](http://www.plantuml.com/plantuml/png/bOwnIWGn48RxFCNFzSkoUG2vqce5jHEHi1dtWZkPa6GByNntavZY40yqnMJu-ORlFwPiOjvvK-dD-M2YOR1uMKvHc93ZJafvoMML07d5h1NAE-DHWblg_nu8vnwEx9Oem_tTmnYSNqkIidtXjARnbeob-2ifLqZo5jMjQWLBU9hBd9u1ap1F-5308bpEoMSSVhWFG0FhFbqdvOAKIdvUodINwN_8z9zbkifQFmp3JJzk4qDqwodNi75jLgYNLEf8tkwPj--5joy0)

First of all lets configure the RMW with the required static memory. The `app-colcon.meta` should looks like:

```
{
    "names": {
        "rmw_microxrcedds": {
            "cmake-args": [
                "-DRMW_UXRCE_MAX_NODES=1",
                "-DRMW_UXRCE_MAX_PUBLISHERS=2",
                "-DRMW_UXRCE_MAX_SUBSCRIPTIONS=2",
                "-DRMW_UXRCE_MAX_SERVICES=0",
                "-DRMW_UXRCE_MAX_CLIENTS=0",
                "-DRMW_UXRCE_MAX_HISTORY=4",
            ]
        }
    }
}
```

On the other hand, `app.c` should looks like the following code:

> ```c

Reviewer: I think that might clarify the tutorial if you set a link to the file example instead of paste the code here.

Author: Ideally you are creating a brand new app, so there is no code on any repo for this Ping Pong app.

<img  width="300" style="padding-right: 25px;" src="https://raw.githubusercontent.com/micro-ROS/micro-ROS.github.io/34c7da536776145693f46e178a2c0f09dcde973e/_docs/tutorials/basic/getting_started_olimex_freertos/4.png">
</div>

## Required hardware

This tutorial uses the following hardware:

| Item | |
|---------------|----------------------------------------------------------|
| Olimex STM32-E407 | [Link](https://www.olimex.com/Products/ARM/ST/STM32-E407/open-source-hardware) |
| Olimex ARM-USB-TINY-H | [Link](https://www.olimex.com/Products/ARM/JTAG/ARM-USB-TINY-H/) |
| USB-Serial Cable Female | [Link](https://www.olimex.com/Products/Components/Cables/USB-Serial-Cable/USB-Serial-Cable-F/) |

Olimex STM32-E407 features a STM32F407 Cortex-M4 microcontroller. It has 1MB Flash and 196 kB RAM of which 64 kB are core coupled (CCM SRAM). Other features of the chip are:

- Three 12 bits @ 2.4MHz ADC
- Two 12 bits DAC
- USB OTG FS
- USB OTG HS
- Ethernet
- 14 timers
- 3 SPI
- 3 I2C
- 2 CAN
- 114 GPIOs

> Most of these peripherals are mapped into Olimex board headers.

Author: Removed but keeping the required hardware in order to have references to the serial-USB cable and JTAG buy links

> The micro-ROS + FreeRTOS port of this evaluation board relies on a STM32CubeMX generated project, so HAL and low-level configuration can be tuned to specific requirements.

Author: Removed because of the last comment

The out-of-the-box HAL configuration included in the port packages configures the minimal communication and debugging peripheral required for micro-ROS.

## Adding a new micro-ROS app

> First of all make sure that you have a **ROS 2 Dashing** installation.

Author: Removed references to ROS 2 distro, it would depend on the `ROS_DISTRO` env var

# Create a workspace and download the micro-ROS tools
mkdir microros_ws 
cd microros_ws
git clone -b $ROS_DISTRO https://github.com/micro-ROS/micro-ros-build.git src/micro-ros-build

# Update dependencies using rosdep
rosdep update
rosdep install --from-path src --ignore-src -y

# Build micro-ROS tools and source them
colcon build
source install/local_setup.bash
```

Now, lets create a firmware workspace that targets all the required code and tools for Olimex development board and FreeRTOS:

```bash
# Create step
ros2 run micro_ros_setup create_firmware_ws.sh freertos olimex-stm32-e407
```

Now you have all the required tools to crosscompile micro-ROS and FreeRTOS for Olimex STM32-E407 development board. At this point, you must know that the micro-ROS build system is a four-step workflow:

1. **Create**: retrieves all the required packages for a specific RTOS and hardware platform.
2. **Configure**: configures the downloaded packages with options such as the micro-ROS application, the selected transport layer or the micro-ROS agent IP address (in network transports).
3. **Build**: generates a binary file ready for being loaded in the hardware.

> 4. **Flash**: load the micro-ROS software in the hardware.

Author: Added  a TODO and starting to write this tutorial

> Micro-ROS apps for Olimex + FreeRTOS are located at `firmware/freertos_apps/apps`. In order create a new application a new folder containing two files: the app code and the RMW configuration.

Reviewer 2: ```suggestion
Micro-ROS apps for Olimex + FreeRTOS are located at `firmware/freertos_apps/apps`. In order to create a new application, create a new folder containing two files: the app code and the RMW configuration.
```

```bash
# Creating a new app
cd firmware/freertos_apps/apps
mkdir my_brand_new_app
cd my_brand_new_app
touch app.c app-colcon.meta
```

> For this example we are going to create a ping pong app where a node sends a ping package with a unique identifier using a publisher and receives the same package a pong subscriber. The node will also answer to 

Reviewer 2: ```suggestion
For this example we are going to create a ping pong app where a node sends a ping package with a unique identifier using a publisher and the same package is received by a pong subscriber. The node will also answer to 
```

![pingpong](http://www.plantuml.com/plantuml/png/ZOwnIWGn48RxFCNFzSkoUG2vqce5jHEHi1dtWZkPa6GByNntavZY10yknMJu-ORlFwPiOjvvK-d3-M2YOR1uMKvHc93ZJafvoMML07d7h1NAE-DPWblg_na8vnwEx9OeZmzFOt1-BK7AzetJciPxCfRYVw1S0SbRLBEg1IpXPIvpUWLCmZpXIm6BS3addt7uQpu0ZQlxT1MK2r0g-7sfqbsbRrVfMrMwgbev3CDTlsqJGtJhATUmSMrMg5TKwaZUxfcttuMt7m00)

> First of all lets configure the RMW with the required static memory. The `app-colcon.meta` should looks like:

Reviewer 2: ```suggestion
First of all lets configure the RMW with the required static memory. The `app-colcon.meta` should look like:
```

> First of all lets configure the RMW with the required static memory. The `app-colcon.meta` should look like:

Author: Done

```
{
    "names": {
        "rmw_microxrcedds": {
            "cmake-args": [
                "-DRMW_UXRCE_MAX_NODES=1",
                "-DRMW_UXRCE_MAX_PUBLISHERS=2",
                "-DRMW_UXRCE_MAX_SUBSCRIPTIONS=2",
                "-DRMW_UXRCE_MAX_SERVICES=0",
                "-DRMW_UXRCE_MAX_CLIENTS=0",
                "-DRMW_UXRCE_MAX_HISTORY=4",
            ]
        }
    }
}
```

> On the other hand, `app.c` should looks like the following code:

Reviewer 2: ```suggestion
On the other hand, `app.c` should look like the following code:
```

> On the other hand, `app.c` should look like the following code:

Author: Done

        rcl_publish(&pong_publisher, (const void*)&rcv_msg, NULL);
      }
    }
    
    usleep(10000);
  } while (true);
}
```

Once the new folder is created, lets configure our new app with a serial transport on UART 3:

```bash
# Configure step
ros2 run micro_ros_setup configure_firmware.sh my_brand_new_app --transport serial --dev 3
```

When the configuring step ends, just build the firmware:

```bash
# Build step
ros2 run micro_ros_setup build_firmware.sh
```

Once the build has successfully ended, lets power and connect the board.  First, connect Olimex ARM-USB-TINY-H JTAG programmer to the board's JTAG port:

<img width="400" style="padding-right: 25px;" src="https://raw.githubusercontent.com/micro-ROS/micro-ROS.github.io/34c7da536776145693f46e178a2c0f09dcde973e/_docs/tutorials/basic/getting_started_olimex_freertos/2.jpg">

Make sure that the board power supply jumper (PWR_SEL) is in the 3-4 position in order to power the board from the JTAG connector:

<img width="400" style="padding-right: 25px;" src="https://raw.githubusercontent.com/micro-ROS/micro-ROS.github.io/34c7da536776145693f46e178a2c0f09dcde973e/_docs/tutorials/basic/getting_started_olimex_freertos/1.jpg">

> You should see the red LED lighting. It time to flash the board:

Reviewer 2: ```suggestion
You should see the red LED lighting. It is time to flash the board:
```

```bash
# Flash step
ros2 run micro_ros_setup flash_firmware.sh
```
## Running the micro-ROS app

The micro-ROS app is ready to connect to a micro-ROS Agent and start talking with the rest of the ROS 2 world.

First of all, create and build a micro-ROS agent:

```bash

> # Download micro-ROS Agent packages

Reviewer 3: ```suggestion
# Download micro-ROS-Agent packages
```

ros2 run micro_ros_setup create_agent_ws.sh

> # Build micro-ROS Agent packages, this may take a while.

Reviewer 3: ```suggestion
# Build micro-ROS-Agent packages, this may take a while.
```

colcon build
```

Then connect the Olimex development board to the computer using the usb to serial cable:

<img width="400" style="padding-right: 25px;" src="https://raw.githubusercontent.com/micro-ROS/micro-ROS.github.io/feature/olimex_freertos_tutorial/_docs/tutorials/basic/getting_started_olimex_freertos/5.jpg">

> ***TIP:** Color codes are applicable to [this cable](https://www.olimex.com/Products/Components/Cables/USB-Serial-Cable/USB-Serial-Cable-F/). Make sure to match Olimex Rx with Cable Tx and viceversa. Rememeber GND!*

Reviewer 3: ```suggestion
***TIP:** Color codes are applicable to [this cable](https://www.olimex.com/Products/Components/Cables/USB-Serial-Cable/USB-Serial-Cable-F/). Make sure to match Olimex Rx with Cable Tx and vice-versa. Remember GND!*
```

Then run the agent:

```bash
# Run a micro-ROS agent
ros2 run micro_ros_agent micro_ros_agent serial --dev [device]
```

***TIP:** you can use this command to find your serial device name: `ls /dev/serial/by-id/*`*

> And finally, lets check that everything is working. We are going to listen to ping topic to check whether the Ping Pong node is publishing its own pings

Reviewer 3: ```suggestion
And finally, let's check that everything is working. We are going to listen to ping topic to check whether the Ping Pong node is publishing its own pings
```

```bash
# Subscribe to micro-ROS ping topic
ros2 topic echo /microROS/ping
```

You should see the topic messages published by the Ping Pong node every 5 seconds:

```
pgarrido@pgarrido:~$ ros2 topic echo /microROS/ping
stamp:
  sec: 20
  nanosec: 867000000
frame_id: '1344887256_1085377743'
---
stamp:
  sec: 25
  nanosec: 942000000
frame_id: '730417256_1085377743'
---
```

> On another command line, lets subscribe to the pong topic

Reviewer 3: ```suggestion
On another command line, let's subscribe to the pong topic
```

```bash
# Subscribe to micro-ROS pong topic
ros2 topic echo /microROS/pong
```

> At this point, we know that our app is publishing pings. Lets check if it also answer to someone else pings:

Reviewer 3: ```suggestion
At this point, we know that our app is publishing pings. Let's check if it also answers to someone else pings:
```

```bash
# Send a fake ping
ros2 topic pub --once /microROS/ping std_msgs/msg/Header '{frame_id: "fake_ping"}'
```

> Now, we should see on the ping subscriber our fake ping along wiht the board pings:

Reviewer 2: ```suggestion
Now, we should see on the ping subscriber our fake ping along with the board pings:
```
